File: clients/java/examples/vectorize-alter-table-add.java
// tag::opening[]
import com.datastax.astra.client.DataAPIClient;
import com.datastax.astra.client.core.vectorize.VectorServiceOptions;
import com.datastax.astra.client.tables.Table;
import com.datastax.astra.client.tables.definition.columns.TableColumnDefinitionVector;
import com.datastax.astra.client.tables.definition.rows.Row;
import com.datastax.astra.client.tables.commands.AlterTableAddColumns;

import java.util.HashMap;
import java.util.Map;

public class Example {

  public static void main(String[] args) {
    // Get an existing table
    Table<Row> table =
        new DataAPIClient("APPLICATION_TOKEN")
            .getDatabase("API_ENDPOINT")
            .getTable("TABLE_NAME");

// end::opening[]

// tag::add-external-provider[]
    // Add a vector column and configure an embedding provider
    AlterTableAddColumns alterOperation = new AlterTableAddColumns()
        // This column will store vector embeddings.
        // The {embedding-provider-name} integration
        // will automatically generate vector embeddings
        // for any text inserted to this column.
        .addColumnVector(
            "VECTOR_COLUMN_NAME",
            new TableColumnDefinitionVector()
                .dimension(MODEL_DIMENSIONS)
                .service(
                    new VectorServiceOptions()
                        .provider("{embedding-provider-name-api}")
                        .modelName("MODEL_NAME")
                        .authentication(Map.of("providerKey", "API_KEY_NAME"))
                )
        )
        // If you want to store the original text
        // in addition to the generated embeddings
        // you must create a separate column.
        .addColumnText("TEXT_COLUMN_NAME");
// end::add-external-provider[]

// tag::add-hugging-face-dedicated[]
    // Define parameters for the embedding provider
    Map<String, Object > params = new HashMap<>();
    params.put("endpointName", "ENDPOINT_NAME");
    params.put("regionName", "REGION");
    params.put("cloudName", "CLOUD_PROVIDER");

    // Add a vector column and configure an embedding provider
    AlterTableAddColumns alterOperation = new AlterTableAddColumns()
        // This column will store vector embeddings.
        // The {embedding-provider-name} integration
        // will automatically generate vector embeddings
        // for any text inserted to this column.
        .addColumnVector(
            "VECTOR_COLUMN_NAME",
            new TableColumnDefinitionVector()
                .dimension(MODEL_DIMENSIONS)
                .service(
                    new VectorServiceOptions()
                        .provider("{embedding-provider-name-api}")
                        .modelName("{embedding-provider-model-name-api}")
                        .authentication(Map.of("providerKey", "API_KEY_NAME"))
                )
        )
        // If you want to store the original text
        // in addition to the generated embeddings
        // you must create a separate column.
        .addColumnText("TEXT_COLUMN_NAME");
// end::add-hugging-face-dedicated[]

// tag::add-openai[]
    // Define parameters for the embedding provider
    Map<String, Object > params = new HashMap<>();
    params.put("organizationId", "ORGANIZATION_ID");
    params.put("projectId", "PROJECT_ID");

    // Add a vector column and configure an embedding provider
    AlterTableAddColumns alterOperation = new AlterTableAddColumns()
        // This column will store vector embeddings.
        // The {embedding-provider-name} integration
        // will automatically generate vector embeddings
        // for any text inserted to this column.
        .addColumnVector(
            "VECTOR_COLUMN_NAME",
            new TableColumnDefinitionVector()
                .dimension(MODEL_DIMENSIONS)
                .service(
                    new VectorServiceOptions()
                        .provider("{embedding-provider-name-api}")
                        .modelName("MODEL_NAME")
                        .authentication(Map.of("providerKey", "API_KEY_NAME"))
                        .parameters(params)
                )
        )
        // If you want to store the original text
        // in addition to the generated embeddings
        // you must create a separate column.
        .addColumnText("TEXT_COLUMN_NAME");
// end::add-openai[]

// tag::add-azure-openai[]
    // Define parameters for the embedding provider
    Map<String, Object > params = new HashMap<>();
    params.put("resourceName", "RESOURCE_NAME");
    params.put("deploymentId", "DEPLOYMENT_ID");

    // Add a vector column and configure an embedding provider
    AlterTableAddColumns alterOperation = new AlterTableAddColumns()
        // This column will store vector embeddings.
        // The {embedding-provider-name} integration
        // will automatically generate vector embeddings
        // for any text inserted to this column.
        .addColumnVector(
            "VECTOR_COLUMN_NAME",
            new TableColumnDefinitionVector()
                .dimension(MODEL_DIMENSIONS)
                .service(
                    new VectorServiceOptions()
                        .provider("{embedding-provider-name-api}")
                        .modelName("MODEL_NAME")
                        .authentication(Map.of("providerKey", "API_KEY_NAME"))
                        .parameters(params)
                )
        )
        // If you want to store the original text
        // in addition to the generated embeddings
        // you must create a separate column.
        .addColumnText("TEXT_COLUMN_NAME");
// end::add-azure-openai[]

// tag::add-hosted-provider[]
    // Add a vector column and configure an embedding provider
    AlterTableAddColumns alterOperation = new AlterTableAddColumns()
        // This column will store vector embeddings.
        // The {embedding-provider-name} integration
        // will automatically generate vector embeddings
        // for any text inserted to this column.
        .addColumnVector(
            "VECTOR_COLUMN_NAME",
            new TableColumnDefinitionVector()
                .service(
                    new VectorServiceOptions()
                        .provider("{embedding-provider-name-api}")
                        .modelName("{embedding-provider-model-name-api}")
                )
        )
        // If you want to store the original text
        // in addition to the generated embeddings
        // you must create a separate column.
        .addColumnText("TEXT_COLUMN_NAME");
// end::add-hosted-provider[]

//tag::closing[]

    table.alter(alterOperation);
  }
}
//end::closing[]

File: clients/typescript/examples/vectorize-alter-table-add.ts
// tag::imports[]
import { DataAPIClient } from "@datastax/astra-db-ts";
// end::imports[]

// tag::init[]

// Get an existing table
const client = new DataAPIClient("APPLICATION_TOKEN");
const database = client.db("API_ENDPOINT");
const table = database.table("TABLE_NAME");
// end::init[]

// tag::add-external-provider[]

// Define the columns and primary key for the table
(async function () {
  await table.alter({
    operation: {
      add: {
        columns: {
          // This column will store vector embeddings.
          // The {embedding-provider-name} integration
          // will automatically generate vector embeddings
          // for any text inserted to this column.
          VECTOR_COLUMN_NAME: {
            type: "vector",
            dimension: MODEL_DIMENSIONS,
            service: {
              provider: '{embedding-provider-name-api}',
              modelName: 'MODEL_NAME',
              authentication: {
                providerKey: 'API_KEY_NAME',
              },
            },
          },
          // If you want to store the original text
          // in addition to the generated embeddings
          // you must create a separate column.
          TEXT_COLUMN_NAME: "text",
        },
      },
    },
  });
})();
// end::add-external-provider[]

// tag::add-hugging-face-dedicated[]

// Define the columns and primary key for the table
(async function () {
  await table.alter({
    operation: {
      add: {
        columns: {
          // This column will store vector embeddings.
          // The {embedding-provider-name} integration
          // will automatically generate vector embeddings
          // for any text inserted to this column.
          VECTOR_COLUMN_NAME: {
            type: "vector",
            dimension: MODEL_DIMENSIONS,
            service: {
              provider: '{embedding-provider-name-api}',
              modelName: '{embedding-provider-model-name-api}',
              authentication: {
                providerKey: 'API_KEY_NAME',
              },
              parameters: {
                endpointName: 'ENDPOINT_NAME',
                regionName: 'REGION',
                cloudName: 'CLOUD_PROVIDER',
              },
            },
          },
          // If you want to store the original text
          // in addition to the generated embeddings
          // you must create a separate column.
          TEXT_COLUMN_NAME: "text",
        },
      },
    },
  });
})();
// end::add-hugging-face-dedicated[]

// tag::add-openai[]

// Define the columns and primary key for the table
(async function () {
  await table.alter({
    operation: {
      add: {
        columns: {
          // This column will store vector embeddings.
          // The {embedding-provider-name} integration
          // will automatically generate vector embeddings
          // for any text inserted to this column.
          VECTOR_COLUMN_NAME: {
            type: "vector",
            dimension: MODEL_DIMENSIONS,
            service: {
              provider: '{embedding-provider-name-api}',
              modelName: 'MODEL_NAME}',
              authentication: {
                providerKey: 'API_KEY_NAME',
              },
              parameters: {
                organizationId: 'ORGANIZATION_ID',
                projectId: 'PROJECT_ID',
              },
            },
          },
          // If you want to store the original text
          // in addition to the generated embeddings
          // you must create a separate column.
          TEXT_COLUMN_NAME: "text",
        },
      },
    },
  });
})();
// end::add-openai[]

// tag::add-azure-openai[]

// Define the columns and primary key for the table
(async function () {
  await table.alter({
    operation: {
      add: {
        columns: {
          // This column will store vector embeddings.
          // The {embedding-provider-name} integration
          // will automatically generate vector embeddings
          // for any text inserted to this column.
          VECTOR_COLUMN_NAME: {
            type: "vector",
            dimension: MODEL_DIMENSIONS,
            service: {
              provider: '{embedding-provider-name-api}',
              modelName: 'MODEL_NAME',
              authentication: {
                providerKey: 'API_KEY_NAME',
              },
              parameters: {
                resourceName: 'RESOURCE_NAME',
                deploymentId: 'DEPLOYMENT_ID',
              },
            },
          },
          // If you want to store the original text
          // in addition to the generated embeddings
          // you must create a separate column.
          TEXT_COLUMN_NAME: "text",
        },
      },
    },
  });
})();
// end::add-azure-openai[]

// tag::add-hosted-provider[]

// Define the columns and primary key for the table
(async function () {
  await table.alter({
    operation: {
      add: {
        columns: {
          // This column will store vector embeddings.
          // The {embedding-provider-name} integration
          // will automatically generate vector embeddings
          // for any text inserted to this column.
          VECTOR_COLUMN_NAME: {
            type: "vector",
            service: {
              provider: '{embedding-provider-name-api}',
              modelName: '{embedding-provider-model-name-api}',
            },
          },
          // If you want to store the original text
          // in addition to the generated embeddings
          // you must create a separate column.
          TEXT_COLUMN_NAME: "text",
        },
      },
    },
  });
})();
// end::add-hosted-provider[]
